src/Corax/Utils/EntryIdEncodings.cs

```diff
@@ -0,0 +1,120 @@
+    private const long Mask = 0xFFL;
+    private const int ContainerTypeOffset = 2;
+
+    
+    
+    private const int AllBuffer = -1;
+    
+    // Quantization parameters:
+    private const long Min = 0;
+    private const long Max = 255;
+    private const long QuantizationMax = 256 - 1;
+    // ReSharper disable once PossibleLossOfFraction
+    private const double QuantizationStep = (Max - Min) / (double)QuantizationMax;
+    
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static long Encode(long entryId, short count, TermIdMask containerType)
+    {
+        Debug.Assert((count & ~Mask) == 0);
+        return count << ContainerTypeOffset | entryId << EntryIdOffset | (long)containerType;
+    }
+
+    /// <summary>
+    /// Returns id of entry without offset and frequency
+    /// </summary>
+    /// <param name="entryId"></param>
+    /// <returns></returns>
+    public static (long EntryId, short Frequency) Decode(long entryId)
+    {
+        return (entryId >> EntryIdOffset, (short)((entryId >> ContainerTypeOffset) & Mask));
+    }
+
+    /// <summary>
+    /// Container id (page) without encodings.
+    /// </summary>
+    public static long GetContainerId(long entryId)
+    {
+        return (entryId >> EntryIdOffset);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static unsafe void Encode(Span<long> entries, ReadOnlySpan<short> frequencies, bool usePtrVersion = false)
+    {
+        Debug.Assert(frequencies.Length == entries.Length);
+
+        if (usePtrVersion == false)
+            goto Classic;
+        
+        fixed (long* entriesPtr = entries)
+        fixed (short* frequenciesPtr = frequencies)
+        {
+            var currentEntryPtr = entriesPtr;
+            var entriesPtrEnd = entriesPtr + entries.Length;
+            var currentFrequency = frequenciesPtr;
+            
+            while (currentEntryPtr != entriesPtrEnd)
+            {
+                *currentEntryPtr = (*currentEntryPtr << EntryIdOffset) | FrequencyQuantization(*currentFrequency);
+
+                currentEntryPtr++;
+                currentFrequency++;
+            }
+
+            return;
+        }
+        
+        Classic:
+        for (int i = 0; i < entries.Length; ++i)
+        {
+            var ecd = FrequencyQuantization(frequencies[i]) << ContainerTypeOffset | entries[i] << EntryIdOffset;
+            entries[i] = ecd;
+        }
+    }
+    
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static long DecodeAndDiscardFrequency(long entryId) => entryId >> EntryIdOffset;
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static void DecodeAndDiscardFrequency(Span<long> entries, int read = AllBuffer)
+    {
+        if (read == AllBuffer)
+            read = entries.Length;
```

Comment:
why add this if, when caller can provide it directly?

Reply:
Leftover, deleted.